### DISCUSSION TITLE
[p.382 6.4.2 FRONT 애플리케이션 배포 502 Bad Gateway]

### DISCUSSION THREAD

**Discussion by Onthesly** [fsoftwareengineer/todo-application#35] [질문과 답변] [answered]
질문을 작성하실 때에는 아래의 내용을 포함 해 주시기 바랍니다.
오류나 에러가 나는 페이지[제목에 작성]
BACKEND 애플리케이션은 OK 표시가 나오는데 FRONT에서는 계속 502만 나옵니다

오류/에러 로그
![image](https://user-images.githubusercontent.com/87744184/146401897-1eac69d3-7bea-418d-ae07-4f05fed668f9.png)
![image](https://user-images.githubusercontent.com/87744184/146402041-3a6b89b2-62fa-4ad4-979b-b953b0b5712e.png)
![image](https://user-images.githubusercontent.com/87744184/146402223-0f820dea-8b5d-41b8-86ae-94ba3fe7dd07.png)

작성 중인 코드
깃허브랑 책이랑 똑같이 작성하였습니다...ㅠ

오류/에러를 해결하기 위해 현재까지 시도 해 본 방법들
1. gitignore 때문일수도 있다해서 .ebignore 작성
2. .elasticbeanstalk/config.yml에 코드추가
```
artifacts:
  files:
    - "**/*"
  base-directory: build
```
3. node 버전 14.15.4로 낮춤(원래 16버전이었습니다.)
4. 환경 속성 강제로 PORT=8080 으로 deploy 해봄.
5. eb terminate하고 build등 삭제후 다시 create해봄.
//////////////////////////////////////////////////////////////////////////////
거의 다 끝났다고 생각했는데 여기서 하루종일 시간날렸네요...방법이 없을까요?ㅠ

**Comment by fsoftwareengineer** [comment]
1. 일라스틱 빈스톡 콘솔에서 로그 파일을 복사해 남겨주시기 바랍니다.
2. 포트 3000은 작동하지 않았나요?

**Reply by Onthesly** [reply]
[TailLogs-1639710441099.txt](https://github.com/fsoftwareengineer/todo-application/files/7731818/TailLogs-1639710441099.txt)


지금도 로컬에서는 잘작동됩니다!!

**Comment by Senna-Lab** [comment]
같은 문제로 헤매는 중인데 혹시 해결하셨나요? ㅠㅠ

**Reply by Onthesly** [reply]
아니요 해결 못했습니다...ㅠ
저는 거의 포기하고 다른 프로젝트 하고 있는데
혹시나 해결 방법 찾으시면 공유 부탁드립니다!!

**Comment by fsoftwareengineer** [comment]
혹시 작성하신 코드를 깃허브에 올리셨다면 확인 해 보도록 하겠습니다.

**Reply by fsoftwareengineer** [reply]
또한 로그 맨 아래에 react-script permission error가 보이므로 아래의 링크를 따라 퍼미션 설정이 가능한지 확인 해 주시기 바랍니다. 

https://forums.aws.amazon.com/thread.jspa?threadID=290964

**Reply by fsoftwareengineer** [reply]
https://stackoverflow.com/questions/49951257/elastic-beanstalk-deployment-failure-permission-denied-mkdir/52836726

.npmrc 파일을 만든 후 unsafe-perm=true를 추가하라는 것 같습니다.

**Reply by Onthesly** [reply]
이런식으로 만들어서 build후 deploy 해봤는데 변함이 없습니다.
![image](https://user-images.githubusercontent.com/87744184/147744054-9dbb7ec7-2387-4a34-bd56-dc50891a969e.png)
제 깃허브 주소입니다.
https://github.com/Onthesly/TodoWebApp
번거롭게 해드려 죄송합니다ㅠ

**Reply by fsoftwareengineer** [reply]
혹시 리액트 다운그레이드 휴 package-lock.json을 삭제 후 다시 디플로이하는 방법이나 package-lock.json을 공식 리파지토리에서 받은걸로 시도 해 보셨나요?

**Reply by fsoftwareengineer** [reply]
npmrc파일은 프로젝트 루트에 존재해야 하는 것 같습니다. 이미 package.json이 있는 경로에 추가한 후 디플로이하는 방법을 시도 하셨나요?

**Comment by Senna-Lab** [comment, marked as answer]
안녕하세요! 에러 해결해서 따로 댓글 올립니다..!

1. elastic beanstalk 환경을 완전 종료 후 react 프로젝트 폴더에서 관련 파일 삭제
2. 처음부터 다시 init, 마지막 단계의 set up SSH 물음에서 Y (이후 설정은 전부 디폴트로 두었습니다.)
3. eb create는 동일하게
4. eb ssh 로 elastic beanstalk의 ec2에 접속합니다.
5. sudo su 로 root 유저로 전환
6. cd ../../ 후 cd /var/app/current (여기에서 ls하면 deploy 했던 파일들이 보여야 합니다)
7. setfacl -R -m u:ec2-user:rwx node_modules (node_modules 및 하위 폴더에 대한 ec2-user의 액세스 권한 부여)
   ==> 로그 확인해보면 이 작업 이후로 sh: /var/app/current/node_modules/.bin/react-scripts: Permission denied 에러가 나지 않았습니다.
8. sudo service nginx restart
9. AWS콘솔 - elastic beanstalk - 구성에서 환경속성 PORT=3000 설정

(혹시 오류나면 
10-1. ps -ef | grep node 로 보고 실행되는 react start가 있으면 kill pid번호로 죽이고 npm run start&
10-2. 혹은 그냥 sudo service nginx restart... )

여기까지 하면 실행이 되었던 것 같습니다..!

**Reply by fsoftwareengineer** [reply]
이 과정을 자동화하는 방법이 있나요? 오토스케일링을 사용하는 경우 새 인스턴스가 실행되거나 하면 항상 작업 해 줘야 할 수 있어서요.

**Reply by Senna-Lab** [reply]
아 그럴 수도 있군요..!! 저는 자동화 까지는 모르겠습니다ㅠㅠ... 

https://stackoverflow.com/a/46001517/769726
여기 작성자 말처럼 EC2 기본 접속 유저를 ec2-user에서 root로 바꾸는 방법이 있다면 더 간편하지 않을까 싶었습니다.. 
(저는 어떻게 해야하는건지 모르겠어서 위의 방법을 선택했습니다... ㅠㅠ)
